subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -5001,4 +5006,185 @@ static int lwm2m_engine_init(const struct device *dev)
 	return 0;
 }
 
+int lwm2m_perform_composite_read_op(struct lwm2m_message *msg, uint16_t content_format,
+				    struct lwm2m_obj_path path_list[], uint8_t path_list_size)
+{
+	struct lwm2m_engine_obj_inst *obj_inst = NULL;
+	int ret = 0;
+	uint8_t num_read = 0U;
+
+	/* set output content-format */
+	ret = coap_append_option_int(msg->out.out_cpkt, COAP_OPTION_CONTENT_FORMAT, content_format);
+	if (ret < 0) {
+		LOG_ERR("Error setting response content-format: %d", ret);
+		return ret;
+	}
+
+	ret = coap_packet_append_payload_marker(msg->out.out_cpkt);
+	if (ret < 0) {
+		LOG_ERR("Error appending payload marker: %d", ret);
+		return ret;
+	}
+
+	/* Add object start mark */
+	engine_put_begin(&msg->out, &msg->path);
+
+	/* Read resource from path */
+	for (int i = 0; i < path_list_size; i++) {
+		/* Copy path to message path */
+		memcpy(&msg->path, &path_list[i], sizeof(struct lwm2m_obj_path));
+
+
+		if (msg->path.level >= 2U) {
+			obj_inst = get_engine_obj_inst(msg->path.obj_id, msg->path.obj_inst_id);
+		} else if (msg->path.level == 1U) {
+			/* find first obj_inst with path's obj_id */
+			obj_inst = next_engine_obj_inst(msg->path.obj_id, -1);
+		}
+
+		if (!obj_inst) {
+			return -ENOENT;
+		}
+
+		ret = lwm2m_perform_read_object_instance(msg, obj_inst, &num_read);
+		if (ret) {
+			return ret;
+		}
+	}
+
+	/* Add object end mark */
+	if (engine_put_end(&msg->out, &msg->path) == 0 &&
+	    lwm2m_validate_engine_end(content_format)) {
+		LOG_ERR("No space for End marker");
+		return -ENOMEM;
+	}
+
+
+	/* did not read anything even if we should have - on single item */
+	if (ret == 0 && num_read == 0U && msg->path.level == 3U) {
+		return -ENOENT;
+	}
+
+	return ret;
+}
+
+static int do_send_op(struct lwm2m_message *msg, uint16_t content_format,
+		      struct lwm2m_obj_path path_list[], uint8_t path_list_size)
+{
+	switch (content_format) {
+#if defined(CONFIG_LWM2M_RW_SENML_JSON_SUPPORT)
+	case LWM2M_FORMAT_APP_SEML_JSON:
+		return do_send_op_senml_json(msg, path_list, path_list_size);
+#endif
+
+	default:
+		LOG_ERR("Unsupported content-format for /dp: %u", content_format);
+		return -ENOMSG;
+	}
+}
+
+static int do_send_reply_cb(const struct coap_packet *response,
+				 struct coap_reply *reply,
+				 const struct sockaddr *from)
+{
+	uint8_t code;
+
+	code = coap_header_get_code(response);
+	LOG_DBG("Send callback (code:%u.%u)",
+		COAP_RESPONSE_CODE_CLASS(code),
+		COAP_RESPONSE_CODE_DETAIL(code));
+
+	if (code == COAP_RESPONSE_CODE_CHANGED) {
+		LOG_INF("Send done!");
+		return 0;
+	}
+
+	LOG_ERR("Failed with code %u.%u. Not Retrying.",
+		COAP_RESPONSE_CODE_CLASS(code), COAP_RESPONSE_CODE_DETAIL(code));
+
+	return 0;
+}
+
+static void do_send_timeout_cb(struct lwm2m_message *msg)
+{
+	LOG_WRN("Send Timeout");
+
+}
+
+int lwm2m_engine_send(struct lwm2m_ctx *ctx, char *path_list[], uint8_t path_list_size)
```

Comment:
I wonder now, perhaps an application should have a chance to register result callback on it's own? So that the application knows whether the data was delivered or not. Up for discussion.

Reply:
Ok good point. Can we re-use some old callback function pointer?

Reply:
I don't see any useful callback type already present for this purpose to be honest.